rust/private/rust_analyzer.bzl

```diff
@@ -0,0 +1,205 @@
+    else:
+        # Compute the number of ../ we need for relative paths.
+        rule_depth = ctx.build_file_path.count("/") + 1
+        crate["is_workspace_member"] = True
+        crate["root_module"] = "../" * rule_depth + info.crate.root.path
+        crate_root = "../" * rule_depth + info.crate.root.dirname + "/../"
+
+    if info.build_info != None:
+        crate["env"].update({"OUT_DIR": ctx.attr.exec_root + "/" + info.build_info.out_dir.path})
+        crate["source"] = {
+            # We have to tell rust-analyzer about our out_dir since it's not under the crate root.
+            "include_dirs": [crate_root, ctx.attr.exec_root + "/" + info.build_info.out_dir.path],
+            "exclude_dirs": [],
+        }
+    crate["env"].update(info.env)
+
+    deps = []
+    for dep in info.deps:
+        deps.append({
+            "name": dep.crate.name,
+            "crate": crate_mapping["ID-" + dep.crate.root.path],
+        })
+
+    crate["deps"] = deps
+    crate["cfg"] = info.cfgs
+    crate["target"] = find_toolchain(ctx).target_triple
+    if info.proc_macro_dylib_path != None:
+        crate["proc_macro_dylib_path"] = ctx.attr.exec_root + "/" + info.proc_macro_dylib_path
+    return "ID-" + info.crate.root.path, crate
+
+# This implementation is incomplete because in order to get rustc env vars we
+# would need to actually execute the build graph and gather the output of
+# cargo_build_script rules. This would require a genrule to actually construct
+# the JSON, rather than being able to build it completly in starlark.
+# TODO(djmarcin): Run the cargo_build_scripts to gather env vars correctly.
+def _rust_project_impl(ctx):
+    rust_toolchain = find_toolchain(ctx)
+    crate_mapping = dict()
+
+    output = dict()
+    output["sysroot_src"] = ctx.attr.exec_root + "/" + rust_toolchain.rust_lib.label.workspace_root + "/lib/rustlib/src/library"
+    output["crates"] = []
+
+    # Gather all crates and their dependencies into an array.
+    # Dependencies are referenced by index, so leaves should come first.
+    idx = 0
+    for target in ctx.attr.targets:
+        if RustAnalyzerInfo not in target:
+            continue
+
+        # Add this crate's transitive deps to the crate mapping and output.
+        for dep_info in target[RustAnalyzerInfo].transitive_deps.to_list():
+            crate_id, crate = create_crate(ctx, dep_info, crate_mapping)
+            if crate_id not in crate_mapping:
+                crate_mapping[crate_id] = idx
+                idx += 1
+                output["crates"].append(crate)
+
+        # Add this crate to the crate mapping and output.
+        crate_id, crate = create_crate(ctx, target[RustAnalyzerInfo], crate_mapping)
+        if crate_id not in crate_mapping:
+            crate_mapping[crate_id] = idx
+            idx += 1
+            output["crates"].append(crate)
+
+    ctx.actions.write(output = ctx.outputs.filename, content = struct(**output).to_json())
```

Comment:
You might want to add a todo or something to replace the `to_json` with https://docs.bazel.build/versions/master/skylark/lib/json.html just to let future authors know about it

Reply:
Added a TODO, it's not in any released version yet